module/zfs/spa.c

```diff
@@ -4588,27 +4589,42 @@ spa_export_common(char *pool, int new_state, nvlist_t **oldconfig,
 	 * reacquire the namespace lock, and see if we can export.
 	 */
 	spa_open_ref(spa, FTAG);
+
+	/*
+	 * Save the txg that is being synced.
+	 */
+	if (spa->spa_sync_on) {
+		sync_txg = spa_syncing_txg(spa);
+	}
+
 	mutex_exit(&spa_namespace_lock);
 	spa_async_suspend(spa);
 	if (spa->spa_zvol_taskq) {
 		zvol_remove_minors(spa, spa_name(spa), B_TRUE);
 		taskq_wait(spa->spa_zvol_taskq);
 	}
-	mutex_enter(&spa_namespace_lock);
-	spa_close(spa, FTAG);
 
-	if (spa->spa_state == POOL_STATE_UNINITIALIZED)
-		goto export_spa;
 	/*
 	 * The pool will be in core if it's openable, in which case we can
 	 * modify its state.  Objsets may be open only because they're dirty,
 	 * so we have to force it to sync before checking spa_refcnt.
+	 *
+	 * If the pool is syncing wait for it to complete. This could take long
+	 * or may never complete if the IO is suspended (spa_suspended = 1).
+	 * Hence, it is best done without holding the spa_namespace_lock to
+	 * avoid blocking other operations which need spa_namespace_lock.
 	 */
-	if (spa->spa_sync_on) {
-		txg_wait_synced(spa->spa_dsl_pool, 0);
+	if (spa->spa_state != POOL_STATE_UNINITIALIZED && spa->spa_sync_on) {
+		txg_wait_synced(spa->spa_dsl_pool, sync_txg + TXG_DEFER_SIZE);
```

Comment:
@sanjeevbagewadi @behlendorf 
The original `txg_wait_synced(spa->spa_dsl_pool, 0);` is equivalent to `tx->tx_open_txg + TXG_DEFER_SIZE`. Is changing to `sync_txg + TXG_DEFER_SIZE` correct?

Reply:
Yes, you are right. The original change was with txg=0. However, since, we drop the namespace_lock I was wondering if more operations could come through into this zpool. Hence, I decided to save away the open_txg when we enter here (line:4597) and used that.

Reply:
I'd prefer not to change this unless we have a clear reason to, let's stick with `txg_wait_synced(spa->spa_dsl_pool, 0);`.

Reply:
Agreed. Reverted that change.